### DISCUSSION TITLE
Component - Scroller

### DISCUSSION THREAD

**Discussion by dawinaj** [ArthurSonzogni/FTXUI#657] [Q&A] [answered]
Please refer to the following code: https://pastebin.com/KAav4EZM
The worker thread iterates through all the instances of Window class, builds windows from their title and contents, to then display in gridbox.

Now in #125 there was shown an implementation of Scroller. Please tell me how can I remake the code in order to somehow use the Scroller inside my WindowTerminal class instead of just returning plain DOM.

I just can't wrap my head around these components, renderers, renders, lambdas, and all this other BS...

I tried this, but it seems to completely break the terminal ;_;
```
Component WindowTerminal::getRenderer()
{
	Elements innards;
	for (const auto& line : lines)
		innards.push_back(line);
	return Scroller(Renderer([&](){ return vbox(innards); }));
}
```

**Comment by ArthurSonzogni** [comment]
Hi @herhor67,

### Memory safety:

In C++, you must ensure yourself pointers/references are still safe to be used. In your example:
```cpp
Component WindowTerminal::getRenderer()
{
	Elements innards;                                           // `innards` are created`
	for (const auto& line : lines)
		innards.push_back(line);
	return Scroller(Renderer([&](){ return vbox(innards); }));
                                                                    // `innards` are deleted`
}
```

So, `WindowTerminal::getRender()` is always unsafe to be used.

### Incorrect use of FTXUI.

`ftxui::Element` are meant to be used only once per frame. You shouldn't store them.
You should store your state like `std::vector<std::string> lines` for instance. 

The amount of things to refactor in the `.cc` file is important. You should probably share the whole repository if you want me to submit PRs.

**Comment by dawinaj** [comment]
Why is it bad to store text nodes instead of creating them every frame? Are they somehow being modified? This sounds like a big waste of memory operations...
I can add you to the repo, but it's private (the entire app is meant for personal use only) so idk if you are willing to contribute to something like that.
Of course you can reuse/repost any code related to the TUI, I just can't share publicly the internal stuff, databases, apis, etc.

**Reply by ArthurSonzogni** [reply]
> Why is it bad to store text nodes instead of creating them every frame?

With FTXUI, I am trying to promote some kind of functional programing with UI. That is to say, instead of mutating the DOM with a sequence of change everytime for all the different operations, you instead define a `state` and a function `state -> UI`. This way everytime "something" changes, it automatically translate into a change in your UI. No more need synchronize your state with your UI state and you have a single source of truth.

> This sounds like a big waste of memory operations...

Yes, but the order of magnitude is so small this doesn't have any consequence in practise.
~ `Premature optimization is the root of all evil`

> I can add you to the repo, but it's private (the entire app is meant for personal use only) so idk if you are willing to contribute to something like that.

I tried for 30min, but I greatly underestimate the difficulty it was using this project. I abandon.

**Comment by dawinaj** [comment]
I understand now what was the problem. After moving the entire `inanrds` and stuff *inside* the renderer lambda, it displays properly. However, scrolling still doesn't work. How should I trigger/event or w/e the Component?

I am now just doing `Element win = ins.getRenderer()->Render();` inside the loop, but I doubt its the correct approach.

**Reply by ArthurSonzogni** [reply, marked as answer]
About making a scrollable area, either you use a the `Scroll` class from [here](https://github.com/ArthurSonzogni/git-tui/blob/master/src/scroller.cpp) at the cost of not being able to have interactive component inside the scrolled area. Alternatively, you use a `Container::Vertical` and plumb some component inside.

**Reply by dawinaj** [reply]
Alright, but can you explain how do I use the Scroller? With the latest commit, it is being displayed properly, but I dont know how to make it receive the scrolling events.
Unless it's too complicated like you said an the answer above, then I'll just stick with a static DOM...
What I mean is, for example, how do I put these Components inside a gridbox, which could be easily done with simple Elements...
